### PR TITLE
ESQL: Add support for table parameters

### PR DESCRIPTION
**Pull request by nik9000**
This adds support for specifying tables in the request for use with the forthcoming `LOOKUP` command. Example:
```
curl -XPOST 'localhost:9200/_query?error_trace&pretty' -d'{
    "query": "FROM foo | LOOKUP services ON hostname | STATS MAX(bytes_out) BY service",
    "tables": {
        "services": {
            "hostname:keyword": [ "a1",   "a2",   "a3",  "b1",  "b2"],
            "service:keyword": ["auth", "auth", "auth", "web", "web"]
        }
    },
    "version": "2024.04.01"
}'
```

The format of these is very experimental. We're not sure at all about the API so I just picked the simplest thing we could merge and iterate on. I went with `name:type` for the keys because I think the types will be important when merging, especially given that we're merging with hashes.

At the moment you can't use this parameter unless you are on a snapshot version. And there's a limit of 1mb for the data in these tables because we, at the moment, can't account for them with our regular `BlockFactory`. Once we can account for them we can raise this limit.

### CODE REVIEW

**Comment by nik9000** [comment]
I've marked this as ES|QL-ui impacting because kibana will likely want to specify these tables. It'll probably want to support user uploaded CSVs and all kinds of other fun things. And this is one way to land the CSV files.

We're going to merge this one without finalizing the syntax just so we can have something to experiment with. So if you see this merged don't think of any decisions about syntax as having been made - just that we're moving towards `LOOKUP` and `INLINESTATS`.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/EsqlFeatures.java [hunk 1]
```diff
@@ -126,6 +126,11 @@ public class EsqlFeatures implements FeatureSpecification {
      */
     public static final NodeFeature METRICS_COUNTER_FIELDS = new NodeFeature("esql.metrics_counter_fields");
 
+    /**
+     * Support for {@code LOOKUP} command
+     */
+    public static final NodeFeature LOOKUP_COMMAND = new NodeFeature("esql.lookup_command");
+
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/EsqlFeatures.java:133]: I cherry-picked this one in from the Lookup PR but didn't plug it in.

File: x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/10_basic.yml [hunk 2]
```diff
@@ -346,3 +346,27 @@ setup:
   - match: {values.0: ["1",2.0,null,true,123,123]}
   - match: {values.1: ["1",2.0,null,true,123,123]}
   - match: {values.2: ["1",2.0,null,true,123,123]}
+
+---
+lookup:
```
Comment (nik9000) [inline x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/esql/10_basic.yml:351]: This test is in the `lookup` PR and seems worth having. It doesn't run at the member, but it illustrates how you'd use this.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlQueryRequest.java [hunk 3]
```diff
@@ -84,11 +94,14 @@ public ActionRequestValidationException validate() {
...
-                "[" + RequestXContent.PRAGMA_FIELD + "] only allowed in snapshot builds",
-                validationException
-            );
+        if (onSnapshotBuild == false) {
+            if (pragmas.isEmpty() == false) {
+                validationException = addValidationError(
+                    "[" + RequestXContent.PRAGMA_FIELD + "] only allowed in snapshot builds",
+                    validationException
+                );
+            }
```
Comment (dnhatn) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlQueryRequest.java:103]: I think we should reject requests with tables in non-snapshot build for now?

**Review by dnhatn — APPROVED** [review]
LGTM. Thanks Nik!

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlQueryRequest.java (hunk 3 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlQueryRequest.java:103]: OH! I totally had that in there at some point. I'll add it back.

**Pull request merged**